### ISSUE TITLE
Refactor container GC logic in order to rely more on database integrity to reduce risk of overly aggressive GCing

### ISSUE THREAD

**Issue by vito** [concourse/concourse#1242]
When we first designed life, we tried to keep the GC logic for containers and volumes as safe and straightforward as possible. "When all the things I pointed to are gone, delete me." The idea behind this design is to have an explicit dependency chain, such that each link has simple GC logic, and Concourse-level semantics like a build finishing and resource checking being done can easily trickle down into the garbage collector, from domain-specific to abstract. Doing it this way reduces the risk of mistakenly killing something that was still in use, as the GC logic is dead simple and relies on the schema and database integrity to prevent bad things from happening. There was no "look at this container, assess the state of the world, and try to decide if we can kill it."

The volume collector has stayed pretty simple. The container collector, however, has not.

Compare the query for finding GC-able volumes:

```sql
SELECT FROM volumes v
LEFT JOIN workers w ON v.worker_name = w.name
WHERE v.initialized
AND v.worker_resource_cache_id IS NULL
AND v.worker_base_resource_type_id IS NULL
AND v.container_id IS NULL
AND w.state IN ('running', 'landing', 'retiring')
```

To the equivalent query for containers:

```sql
SELECT FROM containers c
LEFT JOIN builds b ON b.id = c.build_id
LEFT JOIN volumes v ON v.worker_resource_cache_id = c.worker_resource_cache_id
LEFT JOIN worker_resource_caches wrc ON wrc.id = c.worker_resource_cache_id
LEFT JOIN (SELECT resource_cache_id, count(*) cnt FROM resource_cache_uses GROUP BY resource_cache_id) rcu ON rcu.resource_cache_id = wrc.resource_cache_id
WHERE (
  (c.build_id IS NOT NULL AND b.interceptible = false)
  OR
  (c.best_if_used_by < NOW())
  OR
  (c.build_id IS NULL AND c.resource_config_id IS NULL AND c.worker_resource_cache_id IS NULL)
  OR
  (c.resource_config_id IS NOT NULL AND c.worker_base_resource_type_id IS NULL)
  OR
  (c.worker_resource_cache_id IS NOT NULL AND v.initialized = true)
  OR
  (c.worker_resource_cache_id IS NOT NULL AND rcu.cnt IS NULL)
)
```

The volume collector is just, "if my worker resource cache is gone, my worker base resource type is gone, and my container is gone, kill me". The columns automatically become `NULL` when their dependency goes away via their own lifecycle, which volumes don't have to know about.

The container collector, however, has 6 branches of logic:

1. If my build is no longer interceptible
2. If I've reached my `best_if_used_by` date (only relevant to `check` containers)
3. If my build, resource config, and resource cache ID are all gone
4. If I'm for checking a resource config but the worker base resource type I was created with is gone
5. If I'm for fetching a resource cache and there's an `initialized` volume for that cache already
6. If I'm for fetching a resource cache and that resource cache is no longer used by anyone

Cases 1, 2, and 3 are pretty straightforward. Case 4 seems in want of a `worker_resource_config` table (similar to `worker_resource_cache`), which would then collapse it into case 3. Case 5, however, is super sketchy, and results in killing containers used for fetching a duplicate resource cache, even if it's an active `get` step in a build. This is a big no-no and resulted in some of the failures demonstrated by #144062135. Case 6 is also a little wonky in that it requires three levels of indirection (container -> worker resource cache -> resource cache -> resource cache uses), but it is trying to say "I was created to fetch a resource cache and the thing that required that is done with its job".

I think the mistake we made was tying the containers directly to resource caches or config. Those don't express a lifecycle that's safe enough to assess whether the container can be GCed. Caches and configs are very abstract and will usually outlive these containers. So we can't simply watch for them to go away.

Instead, we should consider the use cases for all containers:

1. `get`, `put`, or `task` in a build.
1. `check`ing of a pipeline resource for up to an hour.
1. `check` and `get` for constructing a container's image (`image_resource` or custom resource type).

And then we can define when to know the container can be removed:

1. When the build is no longer `interceptible`.
1. When the checking has reached its `best_if_used_by` date.
1. When the container that we were fetching the image for has been created.

The important part of having these particular relationships is that these three cases are all things that Concourse will ensure happens eventually:

1. All builds will eventually complete in some way.
1. If nothing else, time will pass and result in us expiring the check session.
1. This one's inherently self-referential; as long as we get container GC right this will eventually go `NULL`. Container creation is idempotent, and we'll re-attach and progress the container to `created` eventually, but if that fails we may want to think about GC logic for stuck `creating` containers if needed.

Here are the proposed schema changes:

* `containers` no longer references `resource_cache`, `resource_config`, `worker_resource_cache`, or `worker_base_resource_type`
* `containers` has a foreign key on `(build_id, build_interceptible=true)`, `ON DELETE SET NULL`. So when the build becomes noninterceptible or the build goes away, this reference is nullified.
* `containers` has a foreign key on `(container_id, container_state='creating')`, `ON DELETE SET NULL`. So when the container state changes to `created` or the container is destroyed, this reference is nullified.
* introduce a `resource_config_check_session` table, which has `id, resource_config_id, worker_base_resource_type, expires_at`. This would be created whenever we're creating a long-lived `check` container, i.e. specifically for checking a resource or resource type.
* `containers` has a foreign key on `resource_config_check_session_id`, `ON DELETE SET NULL`.

This introduces a new table, so we'd need to define the lifecycle for that to fully define container GC:

* created when checking a `resource_config_id` that has no existing check session.
* `resource_config_id` should be `ON DELETE RESTRICT` - we'll flesh this out later in #1247
* `worker_base_resource_type` is `ON DELETE CASCADE`; no GC necessary
* Then, just delete if `expires_at < NOW()`.

**Comment by clarafu** [comment]
We're going to start with having the `check` and `get`s for image resource fetching reference the creating container. The first step was a small refactor to defer the fetching to a point in time where we have a `Creating` container that we can reference.

To do the actual fetching, we added a general `FindOrCreateContainer` alongside the `FindOrCreateBuildContainer` and `CreateGetContainer` methods on container provider. The sibling methods should be removed after this refactor is complete.

The new general method was introduced because the current methods tie their ownership (GC lifecycle) to their usage (what they're doing). This refactor breaks that up, as checking and getting will now either be done by (creating container OR check session) and (creating container OR build), respectively.

This general method will take a `db.ContainerOwner`, which is a very thin abstraction that just determines the columns to set or query for.

Its usage pattern is vaguely similar to `db.ResourceUser`, so at the end of this we may want to see if there's a redundancy there, e.g. if the user and owner can always just point to the same data.

- @vito & @clarafu

**Comment by clarafu** [comment]
One bummer that we ran in to is that a multi-column foreign key only triggers its `ON DELETE` rule if the referenced columns are actually *deleted*, as opposed to changing their value. Given the wording of the rule that kinda makes sense.

Because of this we can't just have containers reference `builds (id, interceptible)` and `containers (id, state)` and have interceptability/state transitions set things to `NULL`. We'll just have to preserve the joining in our query. In the very short-term this means the query will actually get *worse*, but once we stop having containers reference caches/configs, it should be some query with a couple joins (one for `builds`, one for `containers`).

We've gotten this working for `check` containers for `image_resource` / custom resource type checking. Next step is the `get`, and then the `get` for a build, and then the `check` for pipeline resources / resource types.

- @vito & @clarafu

**Comment by clarafu** [comment]
Everything's now using our generic `db.ContainerOwner`, except for check containers.

We ended up having to split `creating_container_id` into `image_check_container_id` and `image_get_container_id`, as these "owners" also double as identifiers, and with both containers simply pointing to the creating container ID, we can't tell the difference. This worked before because one referenced a resource config and the other referenced a resource cache. The split columns was a pretty simple change.

As a result of this we were able to remove `worker_resource_cache_id` from the `columns` table and simplify the garbage collection. We were also able to remove all `FindOrCreateBuildContainer`-like code paths, resulting in a big cleanup. So, now `get`, `put`, and `task` steps are all owned by a build, using the same creation code path, and the `check` and `get` containers are owned by a "checking for container" and "getting for container", respectively, *also* using the same code path as the other containers, just with different owners.

Moving on to check containers, we're exploring how this would look in the `radar`, which is the only place that uses `NewCheckResource`. We're thinking it'll just need to find-or-create a resource config check session, and call the new general `NewResource` method with that as its owner. There's one race here, however: resource config check sessions have an `expires_at`, so there's a chance that it finds a session just before it expires, and then try to create a container referencing it, and then fail. This problem exists in a different form today: we can find a check container just before it reaches `best_if_used_by`, and then the garbage collector could kill it, causing the check to fail.

We're thinking one simple way to solve this is to have the find-or-create immediately create a new session once the old one has expired, but have the garbage-collector wait for a grace period before actually removing it. Something like 5 minutes, just to pick a nice number.

After we do all this, there is further cleanup to be done in the `worker.ContainerProvider` object - now that it doesn't need a bunch of different ways to create containers, it can just have one FindOrCreate method, without having to use the closures to inject the DB find/create functions.

- @vito & @clarafu
